### ISSUE TITLE
Report on Pointer Analysis Issues

### ISSUE THREAD

**Issue by sherrysun0613** [SVF-tools/SVF#1033]
Dear SVF developer:

&emsp;​We are several PhD students and recently do some research on open-source projects. We have used SUPA [1] to analysis them, but found the following issues:

1. *Context sensitivity*. In example `swap.c` [2], we run the following command.

   ```bash
   ./dvf -cxt -query=all -max-cxt=5 -flow-bg=10000 -cxt-bg=10000 -print-query-pts swap.ll
   ```

​&emsp;The expected result should be that variable `a`and `m` point to `p` and `p'` at their own callsites, respectively. But actual results show they point to the same formal parameter `p`, which is       the 0th arg of function `swap`.

​     2. *Virtual function.* From program `main.c` [3], the call chain sequence is as following:

```c++
db->Put(wrt_opt, key, val); // 1st, in src/main.cc
Status Put(const WriteOptions& options, const Slice& key, const Slice& value); // 2nd, include/tinykv.h
Status DBImpl::Put(const Slice& key, const Slice& val); // 3rd, src/db_impl_write.cc
Status KVDB::Put(const Slice& key, const Slice& value); // 4th, src/db_impl_write.cc
```

&emsp;We execute the commands in `README.md`, and then run the following command to generate `ctx` result.

```shell
./dvf -cxt -query=all -max-cxt=5 -flow-bg=10000 -cxt-bg=10000 -print-query-pts bitcode/libTinyKV.a.ll
```

&emsp;The expected result should be that actual parameter `key` at 1st, `key` at 2nd, `key` at 3rd and `key` at 4th are all aliases to each other. But actual result shows that, only `key` at 3rd and      `key` at 4th are aliases.

&emsp;We are not sure if we run SUPA in a wrong manner, or maybe the version after `SVF-2.5` just not deal with them well. And we report them to you and sincerely ask for advice.

&emsp;The SVF team has been working on improving pointer analysis over the years, and we hope these suggestions also have a little help. Look forward to your reply on how to solve these problems. Thank you!

   [1] https://github.com/SVF-tools/SVF/tree/f50cd0758d49d291a51fbdfa25080ea295ff870d/tools/DDA/
   [2] SVF_exp_mail/context_sen_exp/swap.c
   [3] SVF_exp_mail/virtual_func_exp/TinyKV/main.c
   [4] https://figshare.com/articles/dataset/SVF_Pointer_Analysis_Issues/22274539

**Comment by yuleisui** [comment]
Would you be able to simplify both code examples as small as possible? It still looks a bit large.

**Comment by sherrysun0613** [comment]
Thank you for your reply and suggestions.&nbsp;I will email you after the code is reedited.


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "SVF-tools/SVF"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年3月15日(星期三) 下午3:38
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [SVF-tools/SVF] Report on Pointer Analysis Issues (Issue #1033)





 
Would you be able to simplify both code examples as small as possible? It still looks a bit large.

**Comment by sherrysun0613** [comment]
I have minimized TinyKV code to approximate 150 lines, is it small enough?

**Comment by yuleisui** [comment]
Would you help dig a bit more to figure out the issue of the context-sensitive implementation for your first example? It would be good if you could come back with some debug observations and problems you identified.

**Comment by sherrysun0613** [comment]
After finding the issue of context-sensitivity about first example, I traced the edge relationships between variables in its SVFG according to .ll file. Would this be helpful for debug?

**Comment by yuleisui** [comment]
Yes, you can dig it deeper to see what are the problem. You may wish to make a pull request for a fix if you can.
